### PR TITLE
Provide JDBC implementation of OAuth2AuthorizationConsentService

### PR DESCRIPTION
**Pull request by ovidiupopa91**
Add new JDBC implementation of the OAuth2AuthorizationConsentService
Add equals and hashCode methods in OAuth2AuthorizationConsent

Closes gh-313

### CODE REVIEW

**Comment by ovidiupopa91** [comment]
hi @jgrandja / @sjohnr . The implementation is ready for review.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/JdbcOAuth2AuthorizationConsentService.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+			insertAuthorizationConsent(authorizationConsent);
+		} else {
+			updateAuthorizationConsent(authorizationConsent);
+		}
+	}
+
+	private void updateAuthorizationConsent(OAuth2AuthorizationConsent authorizationConsent) {
+		List<SqlParameterValue> parameters = this.authorizationConsentParametersMapper.apply(authorizationConsent);
+		SqlParameterValue registeredClientId = parameters.remove(0);
+		SqlParameterValue principalName = parameters.remove(0);
```
Comment (sjohnr) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/JdbcOAuth2AuthorizationConsentService.java:126]: I see you did this once in the previous PR, and also notice that this `remove`/`add` is done in the `JdbcOAuth2AuthorizedClientService` in spring-projects/spring-security#8435. But is there a specific reason for doing this? I'm unaware of a need for it.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/JdbcOAuth2AuthorizationConsentService.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+		public List<SqlParameterValue> apply(OAuth2AuthorizationConsent authorizationConsent) {
+			List<SqlParameterValue> parameters = new ArrayList<>();
+			parameters.add(new SqlParameterValue(Types.VARCHAR, authorizationConsent.getRegisteredClientId()));
+			parameters.add(new SqlParameterValue(Types.VARCHAR, authorizationConsent.getPrincipalName()));
+
+			Set<String> authorities = new HashSet<>();
+			for (GrantedAuthority authority : authorizationConsent.getAuthorities()) {
+				authorities.add(authority.getAuthority());
+			}
+			parameters.add(new SqlParameterValue(Types.VARCHAR, StringUtils.collectionToDelimitedString(authorities, ",")));
```
Comment (sjohnr) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/JdbcOAuth2AuthorizationConsentService.java:244]: Can this be replaced with:

```java
parameters.add(new SqlParameterValue(Types.VARCHAR, StringUtils.collectionToDelimitedString(authorizationConsent.getAuthorities(), ",")));
```

Any reason to copy over the authorities first?

**Review by sjohnr — COMMENTED** [review]
Thanks for your contribution, @ovidiupopa91! Please see my feedback inline.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/JdbcOAuth2AuthorizationConsentService.java (hunk 2 above)
Comment (sjohnr) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/JdbcOAuth2AuthorizationConsentService.java:244]: Nope, I think what you did makes sense. I was forgetting about the authority field hanging off of `GrantedAuthority`.

**Review by sjohnr — APPROVED** [review]

**Pull request merged**